bookkeeper-server/src/test/java/org/apache/bookkeeper/client/BookKeeperBuildersTest.java

```diff
@@ -0,0 +1,127 @@
+import org.mockito.stubbing.Answer;
+
+/**
+ * Unit tests of builders in this package
+ */
+public class BookKeeperBuildersTest {
+
+    @Test
+    public void testCreateLedgerDefaults() throws Exception {
+
+        int ensembleSize = 3;
+        int writeQuorumSize = 3;
+        int ackQuorumSize = 3;
+        Map<String, byte[]> customMetadata = new HashMap<>();
+
+        BookieWatcher mockWatcher = mock(BookieWatcher.class);
+        when(mockWatcher.newEnsemble(ensembleSize, writeQuorumSize, ackQuorumSize, customMetadata))
+                .thenReturn(new ArrayList<>(Arrays.asList(new BookieSocketAddress("localhost", 1234))));
+
+        NullStatsLogger nullStatsLogger = new NullStatsLogger();
+        BookKeeper bk = mock(BookKeeper.class);
+        when(bk.getCreateOpLogger()).thenReturn(nullStatsLogger.getOpStatsLogger("mock"));
+        when(bk.getCloseLock()).thenReturn(new ReentrantReadWriteLock());
+        when(bk.isClosed()).thenReturn(false);
+        when(bk.getBookieWatcher()).thenReturn(mockWatcher);
+        when(bk.getExplicitLacInterval()).thenReturn(0);
+
+        ClientConfiguration defaultConfig = new ClientConfiguration();
+
+        when(bk.getConf()).thenReturn(defaultConfig);
+        when(bk.getStatsLogger()).thenReturn(nullStatsLogger);
+
+        LedgerIdGenerator ledgerIdGenerator = mock(LedgerIdGenerator.class);
+        when(bk.getLedgerIdGenerator()).thenReturn(ledgerIdGenerator);
+
+        long mockLedgerId = 12342L;
+
+        Mockito.doAnswer((Answer<Void>) new Answer<Void>() {
+            @Override
+            @SuppressWarnings("unchecked")
+            public Void answer(InvocationOnMock invocation) throws Throwable {
+                Object[] args = invocation.getArguments();
+                BookkeeperInternalCallbacks.GenericCallback cb = (BookkeeperInternalCallbacks.GenericCallback) args[0];
+                cb.operationComplete(BKException.Code.OK, mockLedgerId);
+                return null;
+            }
+        }).when(ledgerIdGenerator).generateLedgerId(any());
+
+        LedgerManager mockLedgerManager = mock(LedgerManager.class);
+        AtomicReference<LedgerMetadata> metadataHolder = new AtomicReference<>();
+
+        Mockito.doAnswer((Answer<Void>) new Answer<Void>() {
+            @Override
+            @SuppressWarnings("unchecked")
+            public Void answer(InvocationOnMock invocation) throws Throwable {
+                Object[] args = invocation.getArguments();
+                BookkeeperInternalCallbacks.GenericCallback cb = (BookkeeperInternalCallbacks.GenericCallback) args[2];
+                metadataHolder.set((LedgerMetadata) args[1]);
+                cb.operationComplete(BKException.Code.OK, null);
+                return null;
+            }
+        }).when(mockLedgerManager).createLedgerMetadata(eq(mockLedgerId), any(), any());
+
+        when(bk.getLedgerManager()).thenReturn(mockLedgerManager);
+
+        CreateBuilder builder = new CreateBuilderImpl(bk);
```

Comment:
@sijie I am new to mockito, do you think this is the right way to use it ?
I think that this way makes sense, it is really tight to the current implementation of ops, anyway all the mocks are about classes in this project so it will be feasible to maintain such tests
If so I will create tests in order to cover all the cases

Thank you

Reply:
This looks good to me. You can make this a util class.